src/Data/Group/Free.purs

```diff
@@ -0,0 +1,42 @@
+module Data.Group.Free (Free(..), free, Signed(..)) where
+  
+import Prelude
+
+import Data.Foldable (foldr)
+import Data.Group (class Group)
+import Data.List (List(..), reverse, (:))
+import Data.Monoid (class Monoid)
+
+data Signed a = Positive a | Negative a
+
+instance showSigned :: Show a => Show (Signed a) where
+  show (Positive x) = "+" <> show x
+  show (Negative x) = "-" <> show x
+
+-- Reduce a term of a free group to canonical form.
+canonical :: forall a. Eq a => List (Signed a) -> List (Signed a)
+canonical = foldr cancelOrPush Nil where
+  cancelOrPush x@(Positive x1) y@(Cons (Negative y1) tl) = if x1 == y1 then tl else x : y
+  cancelOrPush x@(Negative x1) y@(Cons (Positive y1) tl) = if x1 == y1 then tl else x : y
+  cancelOrPush x y = x : y
+
+-- The free group generated by elements of a, up to equality.
+newtype Free a = Free (List (Signed a))
```

Comment:
Maybe this should be named FreeGroup, to distinguish from other free things, such as free monads. Currently free monads are also referred to as Free... https://github.com/purescript/purescript-free/blob/master/src/Control/Monad/Free.purs I'm not sure of the best approach.

Reply:
haskell's free group library calls this `Word`; I'd be okay with either `Free` or `FreeGroup`, but `Word` might be a little too mathematical and I'd rather make the connection with other free constructions explicit.